cpp/src/arrow/array/array_test.cc

```diff
@@ -1307,6 +1307,14 @@ TEST(TestBooleanArray, TrueCountFalseCount) {
   CheckArray(checked_cast<const BooleanArray&>(*arr));
   CheckArray(checked_cast<const BooleanArray&>(*arr->Slice(5)));
   CheckArray(checked_cast<const BooleanArray&>(*arr->Slice(0, 0)));
+
+  // GH-41016 true_count() with array without validity buffer with null_count of -1
+  auto data = ArrayFromJSON(boolean(), "[true, false, true]")->data();
+  data->null_count = -1;
+  auto arr_unknown_null_count = std::make_shared<BooleanArray>(data);
```

Comment:
Not sure why we're redoing this after `ArrayFromJSON` already returned an array. Can't you just reuse the original array?

Reply:
Ah, yes of course. For some reason I had in my head I needed to recreate an array object with the changed null count, but I can of course just update the null_count of the the array inplace. Updated.